controlplane/kubeadm/controllers/kubeadm_control_plane_controller_test.go

```diff
@@ -655,3 +667,301 @@ func TestReconcileInitializeControlPlane(t *testing.T) {
+	g := gomega.NewWithT(t)
+
+	fakeClient := fake.NewFakeClientWithScheme(scheme.Scheme)
+
+	m := &clusterv1.Machine{}
+	node, err := getMachineNode(context.Background(), fakeClient, m)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	g.Expect(node).To(gomega.BeNil())
+}
+
+func Test_getMachineNodeNotFound(t *testing.T) {
+	g := gomega.NewWithT(t)
+
+	fakeClient := fake.NewFakeClientWithScheme(scheme.Scheme)
+
+	m := &clusterv1.Machine{
+		Status: clusterv1.MachineStatus{
+			NodeRef: &corev1.ObjectReference{
+				Kind:       "Node",
+				APIVersion: corev1.SchemeGroupVersion.String(),
+				Name:       "notfound",
+			},
+		},
+	}
+	node, err := getMachineNode(context.Background(), fakeClient, m)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	g.Expect(node).To(gomega.BeNil())
+}
+
+func Test_getMachineNodeFound(t *testing.T) {
+	g := gomega.NewWithT(t)
+
+	node := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "testNode",
+		},
+	}
+	fakeClient := fake.NewFakeClientWithScheme(scheme.Scheme, node.DeepCopy())
+
+	m := &clusterv1.Machine{
+		Status: clusterv1.MachineStatus{
+			NodeRef: &corev1.ObjectReference{
+				Kind:       "Node",
+				APIVersion: corev1.SchemeGroupVersion.String(),
+				Name:       "testNode",
+			},
+		},
+	}
+	node, err := getMachineNode(context.Background(), fakeClient, m)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	g.Expect(node).To(gomega.Equal(node))
+}
+
+func TestKubeadmControlPlaneReconciler_updateStatusNoMachines(t *testing.T) {
+	g := gomega.NewWithT(t)
+
+	cluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "foo",
+			Namespace: "test",
+		},
+	}
+
+	kcp := &controlplanev1.KubeadmControlPlane{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Namespace,
+			Name:      "foo",
+		},
+	}
+	kcp.Default()
+	g.Expect(kcp.ValidateCreate()).To(gomega.Succeed())
+
+	g.Expect(clusterv1.AddToScheme(scheme.Scheme)).To(gomega.Succeed())
+	g.Expect(bootstrapv1.AddToScheme(scheme.Scheme)).To(gomega.Succeed())
+	g.Expect(controlplanev1.AddToScheme(scheme.Scheme)).To(gomega.Succeed())
+	fakeClient := fake.NewFakeClientWithScheme(scheme.Scheme, kcp, cluster)
+	log.SetLogger(klogr.New())
+
+	r := &KubeadmControlPlaneReconciler{
+		Client:                    fakeClient,
+		Log:                       log.Log,
+		TemplateCloner:            &templateCloner{},
+		KubeadmConfigGenerator:    &kubeadmConfigGenerator{},
+		MachineGenerator:          &machineGenerator{},
+		RemoteClusterClientGetter: &remoteClientGetter{},
+	}
+
+	g.Expect(r.updateStatus(context.Background(), kcp, cluster)).To(gomega.Succeed())
+	g.Expect(kcp.Status.Replicas).To(gomega.BeEquivalentTo(0))
+	g.Expect(kcp.Status.ReadyReplicas).To(gomega.BeEquivalentTo(0))
+	g.Expect(kcp.Status.UnavailableReplicas).To(gomega.BeEquivalentTo(0))
+	g.Expect(kcp.Status.Initialized).To(gomega.BeFalse())
+	g.Expect(kcp.Status.Ready).To(gomega.BeFalse())
+	g.Expect(kcp.Status.Selector).NotTo(gomega.BeEmpty())
```

Comment:
Should we try to match the actual selector value?

Reply:
We can, but since we are setting the values programmatically (and it isn't based on user input), I didn't see a need to either hardcode an expected string here for the tests or replicate the logic in the controller to create the expected string here.

Reply:
WFM